tools/codegen/api/autograd.py

```diff
@@ -335,9 +335,34 @@ def repl(m: Match[str]) -> str:
                     required_primals = required_primals + ("self",) if required_primals else ("self",)
 
                 if not is_exact_match:
-                    # Make sure that the forward grad is modified inplace when the original formula
-                    # is out of place
-                    formula = f"self_t_raw.defined() ? self_t_raw.copy_({formula}) : {formula}"
+                    # NOTE [In-place forward AD formula Optimization]
+                    #
+                    # This optimization transforms the formula to directly do inplace, i.e.
+                    # instead of self_t.copy_(self_t.op()) we do self_t.op_() when the following are met:
+                    #
+                    # 1) the formula satisfies the pattern: "self_t.op(*args)"
+                    # 2) "op" in (1) needs to be the same as the op the derivative is for
+                    #
+                    # (2) may seem too strict, but currently the only ops that satisfy (1) also satisfy (2)
+                    # If there is a need, we can relax (2) to allow any op that has an in-place variant
+                    # We can also avoid hard coding self_t, if there is a need, but that is more work.
+                    is_single_method_on_self_t = False
+                    match = re.fullmatch(r'self_t.([\w]*)\((.*)\)', formula)
+                    if match:
+                        op_name, between_parens = match.group(1), match.group(2)
+                        # We want to...
+                        #   Match: self_t.op1(other_p.op2(arg))
+                        #   Avoid: self_ self_t.op1(args) + self_t.op2(args)
+                        # We check this by making sure the first parens in second group is not a closing parens
```

Comment:
This won't be necessary once you update the regex above

Reply:
The issue is that I used `.*` in the second group, so something like  `args) + self_t.op2(args` could match. 

One alternative method to what I'm currently doing is using `[^\(\)]*` as the pattern for the second group, but that would prevent us from matching something with nested parenthesis like `self_t.op1(other_p.to(self_p.dtype))`.

Reply:
The thing is that the check done here will fail in that case as well if you have `self_t.op1(other_p.to(self_p.dtype)) + self_t.op2(args)`

I feel like the simplest way to do this here is to just manually traverse the string in a for loop and check that parenthesis are balanced.

Reply:
Good catch, updated!